### DISCUSSION TITLE
Unable to configure OpenCL with Imagemagick

### DISCUSSION THREAD

**Discussion by blackbird-47** [ImageMagick/ImageMagick#2545] [Help] [answered]
We are trying to configure opencl with imagemagick v7.0.10. We are using an NVidia GPU. 
During configure we specify --enable-opencl flag, but it doesn't recognize it (see screenshot). 
![Screenshot 2020-09-10 at 6 20 35 PM](https://user-images.githubusercontent.com/25073351/92731301-84fa6a80-f392-11ea-9064-c0b940cb3b2c.png)
Although in the configure logs it does show OpenCL as one of the features (see screenshot).  
![Screenshot 2020-09-10 at 6 20 14 PM](https://user-images.githubusercontent.com/25073351/92731332-8e83d280-f392-11ea-96b2-f1bdcbf3ce86.png)
Once make install is complete, when we check the imagemagick version with `convert -version` it doesn't show OpenCl enabled. It seems weird as during configure we did see OpenCL.
![Screenshot 2020-09-10 at 6 19 43 PM](https://user-images.githubusercontent.com/25073351/92731359-9479b380-f392-11ea-8e14-3336fe113462.png)
![Screenshot 2020-09-10 at 6 20 42 PM](https://user-images.githubusercontent.com/25073351/92731224-701dd700-f392-11ea-8f69-1a40c405d0df.png)

Upon running resize operator (which is openCL enabled, should use GPU) we didn't find any device score info in $HOME/.cache/ImageMagick.

Is there any configuration being missed here that is leading to opencl not getting enabled?
GPU: Nvidia and corresponding drivers are installed.
Imagemagick version: 7.0.10-29
OpenCL library installed: ocl-icd-opencl-dev
Configure command: `./configure --enable-opencl --enable-hdri --enable-openmp LIBS=-ldl`

**Comment by dlemstra** [comment, marked as answer]
There is no OpenCL support for Nvidia graphics cards. There were a lot of issues with those cards so we had to disable this because it kept crashing for unknown reasons.

**Reply by blackbird-47** [reply]
Could you please provide which GPUs are supported at present? Thanks

**Reply by dlemstra** [reply]
We only explicitly disabled NVidia cards.

**Reply by bliskner** [reply]
Hi dlemstra,
some more obvious hint during the ./configure step that nvidia cards are currently disabled would be highly appreciated (would saved me a lot of time).
Can you provide any further information on how it could be reactivated for testing purposes? And/or which was the last version that was still capable (althought broken)?
Best

**Reply by dlemstra** [reply]
The disabling is done inside code so we cannot report that during the `./configure` step. You can enable NVIDIA cards if you remove the following code block: https://github.com/ImageMagick/ImageMagick/blob/60c07a8bae6dc39c8155bec4094ff354e350622d/MagickCore/opencl.c#L2440-L2445

**Reply by irisjae** [reply]
Hi @dlemstra, thank you for sharing your knowledge! Are any other modifications to the code necessary to undisable NVIDIA cards? I have removed the mentioned code block, alas `magick -version` after `./configure --enable-opencl; make; sudo make install;` only gives me
```
Version: ImageMagick 7.0.10-34 Q16 x86_64 2020-10-30 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5)
Delegates (built-in): bzlib fontconfig freetype jng jpeg ltdl lzma png x zlib
```

**Reply by WolfBullet** [reply]
The windows install with opencl still uses nvidia gpu's. I tested this on my host that's why I know but I have the same issue in my docker container. If that would be in the docs of imagemagick that would've saved us a lot of time.... But @irisjae i also still have the same in the build it looks good but --version doesn't show opencl

**Reply by Pupa-git** [reply]
> The disabling is done inside code so we cannot report that during the `./configure` step. You can enable NVIDIA cards if you remove the following code block:
> 
> https://github.com/ImageMagick/ImageMagick/blob/60c07a8bae6dc39c8155bec4094ff354e350622d/MagickCore/opencl.c#L2440-L2445

Hello. I have a problem when assembling Imagemagick 6.9.12-4 using OpenCL on a Nvidia video card

I attach the output of commands before and after make install
Perhaps you can tell me what the problem?

![Снимок экрана от 2021-03-23 20-39-39](https://user-images.githubusercontent.com/74722221/112193790-4daa2780-8c19-11eb-90c6-9d7faa129777.png)
![Снимок экрана от 2021-03-23 20-40-05](https://user-images.githubusercontent.com/74722221/112193793-4e42be00-8c19-11eb-91ff-3a7f18d83aa4.png)
![Снимок экрана от 2021-03-23 20-49-10](https://user-images.githubusercontent.com/74722221/112193795-4edb5480-8c19-11eb-928d-7fdde7640701.png)

**Comment by duerra** [comment]
Is there any momentum towards re-enabling this at some point?  After 4 years, perhaps any bugs have been fixed?

**Comment by dlemstra** [comment]
Recent versions now have support for NVIDIA but OPENCL has been disabled by default. You can enable it by setting the environment variable `MAGICK_OCL_DEVICE` to `GPU` or `CPU`.

**Reply by fmw42** [reply]
Dirk, Is OpenCL disable for all graphics cards or only for NVIDIA?

**Reply by dlemstra** [reply]
It is disabled for all graphic cards, it was changed to opt-in.
